Associated constant assignment versus equality

zygoloid opened the pull request:
Split the `where A == B` constraint in two: `where .A = B` produces a new
constraint from an old one, where the value of `.A` in the new constraint is
known and eagerly rewritten to `B`, and `where A == B`, which does not cause
`A` and `B` to be considered as identical by language rules but does permit
implicit (no-op) conversion between them.
 
This aims to provide an efficiently-computable and human-understandable type
equality rule, with type canonicalization and therefore transitive type
equality, without sacrificing too much in the way of ergonomics and without
sacrificing determinism, while still providing the full power of a general type
constraint system in a less ergonomic form.

zygoloid committed:
Proposal: associated constant assignment versus equality

zygoloid committed:
Fill in remaining sections.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+    whether and how we might restrict constraints in order to give a decidable
+    type equality rule. #818 addressed that question by using a non-transitive
+    rule with `observe` declarations to close the transitivity gaps.
+-   [#2070: always `==` not `=` in `where` clauses](https://github.com/carbon-language/carbon-lang/pull/2070)
```
Clarify that #2070 has not been accepted/merged, and this proposal replaces it?

zygoloid replied:
Done.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+-   [#2070: always `==` not `=` in `where` clauses](https://github.com/carbon-language/carbon-lang/pull/2070)
+    describes our previous approach to the "equal types with different
+    interfaces" problem.
+
```
It would seem nice to also mention the status quo without #2070, where `=` has a meaning similar to the meaning here, but is restricted to the case where the right-hand side is a concrete type.

zygoloid replied:
Done, corrections would be appreciated.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+  let Slice:! Container where .Element = Element;
+}
+// `T.Slice.Element` rewritten to `T.Element` because type of `T.Slice` says `.Element = Element`.
+// `T.Element` rewritten to `i32` because type of `T` says `.Element = i32`.
```
Please wrap to fewer columns.

zygoloid replied:
```suggestion
// `T.Slice.Element` rewritten to `T.Element`
//     because type of `T.Slice` says `.Element = Element`.
// `T.Element` rewritten to `i32`
//     because type of `T` says `.Element = i32`.
```

josh11b commented on proposals/p2173.md:
````diff
@@ -0,0 +1,489 @@
+fn A[T:! Container where .Element = i32](x: T.Slice.Element) {}
+```
+
+Rewrites aren't performed on the left-hand side of such an `=`, so
+`where .A = .B and .A = C` is not rewritten to `where .A = .B and .B = C`.
````
Is there a situation like `where .A = .B and .A.C = .D` where we might want to rewrite that to `where .A = .B and .B.C = .D` or is  the left-hand side restricted to a single `.`? If so, might be nice to say that explicitly or with an example.

zygoloid replied:
I'm not sure whether we might want things like that. I'm having a hard time finding a reasonable example constraint where a rewrite applied to the LHS of another rewrite would turn it into a `.X = ` rewrite. `where .A = .Self and .A.B = ...`, or `let X:! C where .A = Self and .A.X.B = T` would have that effect but neither seems especially likely to appear. For now I'm inclined to be restrictive even if we could rewrite to something that would then follow the rules.

The opening sentence of this section ("In a rewrite constraint, the left-hand operand of `=` must be the dotted name of an associated constant.") was intended to convey the "restricted to a single `.`" rule. Maybe "must be a `.` followed by the name of an associated constant" would be clearer. Will make that change.

TODO: Add an example.

zygoloid replied:
Done.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+Replace the existing `where A == B` constraint with two different kinds of
+constraint:
+
+-   `where .A = T` specifies a _rewrite_ constraint. If the associated constant
```
Can `.A` be `.Self`? Might be good to have that as an example.

zygoloid replied:
That might not be problematic, but I'm not sure whether it's useful and it would complicate things. Because rewrites change the type as well as the value, a `.Self` rewrite would make the resulting value effectively be an alias, so I think we should disallow it.

TODO: Add an example demonstrating this is not allowed.

zygoloid replied:
Made the rule more explicit and added examples.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+they are different types that have an isomorphic set of values, somewhat like
+`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
+type, it is requesting different behavior.
+
```
Would it be useful to have some dedicated sections with examples to clarify the rules? Here are some that I think illustrate some important points:

A variation on the example above:
```
interface Allowed1 {
  let A:! Allowed1 where .A = .Self;
}
fn F1(T:! Allowed1, x: T.A.A.A);
// Allowed, since rewrite is only applied once per sub-expression
// T.A.A.A = (T.A.A).A -> (T.A).A = T.A.A -> T.A
```

`Allowed1` with the sides switched leads to errors (similar error as your second `Container` example).
```
interface Error1 {
  let E:! Error1 where .Self = .E;
}
fn F2(T:! Error1, x: T.E);
// T.E -> T.E.E -> T.E.E.E error, rewrite applied twice.
// Clearly can't keep applying this rewrite.
// Heuristic: put the "shorter" expression on the right.
```
This is a case we could diagnose with just the interface definition, since the right-hand side of the = already matches the rewrite rule.

Two rewrites create a loop:
```
interface Example2 {
  let B:! Example2;
  let C:! Example2 where .Self = B;
}
// `where .Self = ...` by itself is fine. `
fn AllowedEx2(T:! Error2, x: T.C);
// But combined with another rewrite, creates an infinite loop.
fn ErrorEx2(T:! Error2 where .B = .C, x: T.C);
// T.C -> T.B -> T.C -> T.B, error: applies T.C -> T.B twice.
// Using `==` instead of `=` would make this redundant, not an error.
```

Rewrite applies twice, but doesn't create an infinite loop
```
constraint ForwardDeclaredConstraint;
interface Error3 {
  let X:! Error3;
  // Means `Y:! Error3 where .X = X.Y`
  let Y:! ForwardDeclaredConstraint;
}
constraint ForwardDeclaredConstraint {
  extends Error3 where .X = X.Y;
}
fn F4[(T:! Error3, x: T.Y.Y.X);
// T.Y.Y.X -> T.Y.X.Y -> T.X.Y.Y
// An error since it applies the same rewrite twice, even
// though there is no infinite loop.
```
This is in some sense a "false positive" of our policy, and the rules could be modified to allow this if this comes up in real code.

zygoloid replied:
I guess that final example needs to be:
```
interface Error3;
constraint ForwardDeclaredConstraint(X:! Error3);
interface Error3 {
  let X:! Error3;
  // Means `Y:! Error3 where .X = X.Y`
  let Y:! ForwardDeclaredConstraint(X);
}
constraint ForwardDeclaredConstraint(X:! Error3) {
  extends Error3 where .X = X.Y;
}
fn F4(T:! Error3, x: T.Y.Y.X);
// T.Y.Y.X -> T.Y.X.Y -> T.X.Y.Y
// An error since it applies the same rewrite twice, even
// though there is no infinite loop.
```

TODO: Add these examples.

josh11b replied:
To no longer use `where .Self =` in `Example2`, can instead say:
```
interface Helper2 {
  let D:! Type;
}
interface Example2 {
  let B:! Type;
  let C:! Helper2 where .D = B;
}
// `where .D = ...` by itself is fine.
fn AllowedEx2(T:! Example2, x: T.C.D);
// But combined with another rewrite, creates an infinite loop.
fn ErrorEx2(T:! Example2 where .B = .C.D, x: T.C.D);
// T.C.D -> T.B -> T.C.D -> T.B, error: applies T.C.D -> T.B twice.
// Using `==` instead of `=` would make this redundant, not an error.
```

josh11b replied:
Rewriting `Error1` to avoid `where .Self =`:
```
interface Error1a {
  let E:! Error1a;
}
constraint Error1b {
  extends Error1a where .E = .E.E;
}
```
(We also discussed:
```
constraint Error1b {
  extends Error1a where .E = .E.E and .E is Error1b;
}
```
but that wasn't necessary and raises other questions.)

zygoloid replied:
Added examples.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+    language rules for any purpose, but there is an `ImplicitAs(T)` impl for
+    types implementing `SameAs(T)`.
+
+A type implements `C where .A = T` if and only if it implements
+`C where .A == T` – the inhabitants of these two type-of-types are the same, but
```
Technically these two conditions are equivalent only when the first one is legal.

zygoloid replied:
Yes, I agree. Which cases are in scope there? This is specifically talking about `where .A == T`, not `where U == T`, so I suppose if `C` already has a rewrite of `.A` to something else then the latter would be valid but not the former. Are there other cases?

zygoloid replied:
I guess it's not important to enumerate how that can happen. Added "assuming both constraints are valid".

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+an associated constant. This notation is permitted anywhere a constraint can be
+written, and results in a new constraint with a different interface: the named
+member effectively no longer names an associated constant of the constrained
+type, and is instead treated as a rewrite rule the expands to the right-hand
```
```suggestion
type, and is instead treated as a rewrite rule that expands to the right-hand
```

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+they are different types that have an isomorphic set of values, somewhat like
+`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
+type, it is requesting different behavior.
+
```
More notes:
- _Uses_ are error, not sets of constraints (not ideal, but may be necessary to make this efficient for the compiler)
- Interaction between `=` and `is`
- Explicitly an experiment to see when `==` is needed. Recommend using `=` over `==` where possible.
- An `=` constraint identified with the source code that defines it, for purposes of defining whether a constraint is reused. Different parameters for interface does not mean different `where` clause.

josh11b reviewed (commented):
Here are some initial comments. I'm still reading and pondering, so I may have more later.

zygoloid committed:
Fix typo.

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

zygoloid commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+### Rewrite constraints
+
+In a rewrite constraint, the left-hand operand of `=` must be the dotted name of
+an associated constant. This notation is permitted anywhere a constraint can be
```
```suggestion
In a rewrite constraint, the left-hand operand of `=` must be a `.` followed by
the name of an associated constant. `.Self` is not permitted.
This notation is permitted anywhere a constraint can be
```

zygoloid commented on proposals/p2173.md:
````diff
@@ -0,0 +1,489 @@
+fn A[T:! Container](x: T.Slice.Element) {}
+```
+
+It is valid for the same rewrite rule to be used multiple times within a type
````
````suggestion
Here, "same rewrite" is determined by whether the same associated constant in
the same interface member is rewritten, ignoring any parameterization of the
interface.

```
interface ContainerOf(T:! Type) {
  let Iterator:! Type;
  // (2), bad constraint
  let Pointers:! ContainerOf(T*) where .Iterator = .Pointers.Iterator;
}
// Error, `T.Pointers.Iterator` rewritten to
// `T.Pointers.Pointers.Iterator` by (2) in `ContainerOf(i32)`, then to
// `T.Pointers.Pointers.Pointers.Iterator` by (2) in `ContainerOf(i32*)`
fn B[T:! ContainerOf(i32)](x: T.Pointers.Iterator) {}
```

It is valid for the same rewrite rule to be used multiple times within a type
````

zygoloid replied:
Is this sufficient to avoid circularity? Can we repeatedly reach constraints applying to somewhere else, such as in a function signature or the parameters of a parameterized type?

The goal here is to avoid the possibility that textually duplicating a constraint would lead to more permissive behavior and to instead behave as if all possible textual duplication of reused constraints had already been performed.

zygoloid replied:
New idea: use the location of the `=` and also the location of the `:!` if any.

zygoloid commented on proposals/p2173.md:
```diff
@@ -0,0 +1,489 @@
+interface Container {
+  let Element:! Type;
+  // (1), bad constraint
+  let Slice:! Container where .Element = .Slice.Element;
```
This is a bad example: lookup presumably fails here because  `Container` is  incomplete.

zygoloid replied:
Fixed.

zygoloid committed:
Apply suggestions from code review

zygoloid committed:
Expand and add examples.

zygoloid committed:
Add some more considered alternatives.

zygoloid committed:
Presubmit.

josh11b commented on proposals/p2173.md:
```diff
@@ -169,11 +171,23 @@ There has been a lot of work in this space. Some previous relevant proposals:
+        left-hand side is not restricted. The type-of-type of the left-hand side
+        is changed to match the right-hand side.
+    -   `==` constraints do not affect the type-of-type of either operand.
+    -   As an ansewr to the question of how to produce a decidable type equality
```
```suggestion
    -   As an answer to the question of how to produce a decidable type equality
```

josh11b commented on proposals/p2173.md:
```diff
@@ -195,33 +209,55 @@ constraint:
 //     because type of `T` says `.Element = i32`.
-fn A[T:! Container where .Element = i32](x: T.Slice.Element) {}
+fn Add[T:! Container where .Element = i32](p: T*, y: T.Slice.Element) {
+  // ✅ Argument `y` has the same type `T.Element` as parameter `x`.
```
```suggestion
  // ✅ Argument `y` has the same type `T.Element` as parameter `x` of
  // `Container.Add`.
```

josh11b commented on proposals/p2173.md:
```diff
@@ -237,38 +273,76 @@ is performed again, the program is invalid.
-// `T.Slice.Slice.Element` by (1), then to
-// `T.Slice.Slice.Slice.Element` by (1)
-fn A[T:! Container](x: T.Slice.Element) {}
+// ❌ Error: `.Slice.Element` rewritten to `.Element` by (1),
```
"(1)" comment was deleted.

josh11b commented on proposals/p2173.md:
````diff
@@ -237,38 +273,76 @@ is performed again, the program is invalid.
+
+```
+interface FalsePositive;
+constraint ForwardDeclaredConstraint(X:! Error3);
````
```suggestion
constraint ForwardDeclaredConstraint(X:! FalsePositive);
```

josh11b commented on proposals/p2173.md:
```diff
@@ -445,6 +519,56 @@ of how we would do this have not been worked out, but this might lead to a
 coherent rule that has transitive type equality. However, this would not solve
 the "equal types with different interfaces" problem.
 
+### Find a fully transitive approach to type equality
```
I think the thing different about this alternative is that it would have transitive _symmetric_ equality, as opposed to requiring the user to specify the direction of rewrite constraints, and without the rewrite restrictions of only allowing one rewrite to apply.

zygoloid replied:
Added some more words about those restrictions.

josh11b reviewed (changes requested):
I've tried to resolve stuff so it is clear what is left.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,616 @@
+the same rewrite.
+
+When an `=` appears within an interface, it is possible to form a cycle where a
+type becomes more complex with each iteration. In order to avoid this problem,
```
Do we have any examples where it becomes more complex with each iteration?

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,616 @@
+these two type-of-types are the same, but they are different type-of-types and
+provide different interfaces for corresponding type values.
+
+## Details
```
Are these points anywhere in the proposal now?

- The "use same rewrite rule twice" error is triggered by invalid type expressions, not invalid constraint definitions. It would be ideal if the "rewrite ends up applying to itself after other rewrites have been applied" rule was sufficient, so errors were always diagnosed as early as possible, but this may not be possible while maintaining efficiency.
- Defining the interaction between `=` and `is` constraints. It isn't clear whether and when `is` constraints are rewritten by `=` rewrites. (Does it matter the order of the `is` constraint and the `=` rewrite? Making them mutually exclusive doesn't seem viable. I think the answer is that `is` constraints are still enforced as requirements but don't affect name lookup, except maybe possibly when they apply to the type of the ultimate destination of rewrites.)
- I think this proposal should explicitly state how this is an experiment to see when `==` is needed, and recommend using `=` over `==` where possible.

zygoloid replied:
I think these points are all now addressed.

josh11b commented on proposals/p2173.md:
```diff
@@ -0,0 +1,616 @@
+Replace the existing `where A == B` constraint with two different kinds of
+constraint:
+
+-   `where .A = T` specifies a _rewrite_ constraint. If the associated constant
```
Concern: `=` can be confusing with `let` initializers and defaults. Consider other operators?

zygoloid replied:
I agree that this can lead to some undesirable syntactic forms (`let T:! Foo where .A = i32 = i32`). Added some "Alternatives considered" text, but I don't have a concrete preferable alternative so I'm not planning on addressing this problem in this proposal. This is also not a new problem with this proposal, so I think we shouldn't couple this proposal to solving this problem.

zygoloid committed:
Respond to review comments.

zygoloid committed:
Add precise rules and termination argument.

Remove the "recursively uses the same rewrite" rule which appears to be
unnecessary to guarantee termination.

josh11b commented on proposals/p2173.md:
```diff
@@ -631,3 +1025,107 @@ usage same-type constraints will have, compared to rewrite constraints, so it's
 would be reasonable to revisit this decision when we have more usage
 information.
+
+### Topological ordering for `=` in `and`
```
Is there no way to buffer constraints until a type is formed, perform the sorting, and then do rewriting, without making things complicated?

zygoloid replied:
There is! That's what we do now.

josh11b commented on proposals/p2173.md:
```diff
@@ -266,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+    associated constant `Element`, the constraint
+    `Container & (Iterable where .Element = i32)` is the same as the constraint
+    `(Container & Iterable) where .Element = i32` which is the same as the
+    constraint `Container where .Element = i32`.
```
This section doesn't address `C where .T = i32 & C where .U = .T`. I would like that to use topological sorting of constraints to make that work. This not working is going to a pitfall to using `=` constraints that is going to cause problems for users of a library that employs `=` constraints.

zygoloid replied:
This now addressed later in the proposal. I found a nice way to allow things like this, and the implementation in explorer seems to work well.

josh11b commented on proposals/p2173.md:
```diff
@@ -631,3 +1025,107 @@ usage same-type constraints will have, compared to rewrite constraints, so it's
+
+### Multi-constraint `where` clauses
+
+Instead of treating `A where B and C` as `(A where B) where C`, we could model
+it as `(A where B) & (A where C)`.
```
I think this would be good property and a big simplification. I agree it is contingent on not rejecting `C where .T = i32 & C where .U = .T`, which motivates solving that problem with a topological ordering.

zygoloid replied:
I think it's desirable for earlier constraints to be in scope when type-checking later constraints. I've added a longer description to explain why I think that.

josh11b commented on proposals/p2173.md:
````diff
@@ -266,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+```
+// ❌ Rewritten member `.U` is mentioned in a context
+// where its rewrite is not in scope.
+fn BadG[A:! C where .T = .U and .U = i32]() {}
````
I would argue this is a more natural order to write these constraints. Consider if rewrite rules were written with an asymmetric operator like `~>` instead of `=`, then this constraint would be written `A:! C where .T ~> .U and .U ~> i32`, which is suggestive of the perhaps even more natural `A:! C where .T ~> .U ~> i32`.

zygoloid replied:
This case is now supported.

josh11b commented on proposals/p2173.md:
```diff
@@ -265,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+interface HasAssoc {
+  let Assoc:! Type;
+}
+// The fully-elaborated declared type of `T` is
```
```suggestion
// The fully elaborated declared type of `T` is
```

josh11b commented on proposals/p2173.md:
```diff
@@ -614,3 +1025,107 @@ usage same-type constraints will have, compared to rewrite constraints, so it's
+`.V` in `.U.V` will fail when checked with `.U` of type `Type`, even though it
+might succeed later if `i32` has a member of type `V`. And worse, for
+`C where .T = F(.Self) and .U = G(.Self)` for arbitrary symbolically-evaluatable
+functions `F` and `G`, whether `.T` refers to `.U` or the other way around can't
```
What sorts of problems could come from the definitions of `F` and `G`?

josh11b commented on proposals/p2173.md:
```diff
@@ -614,3 +1025,107 @@ usage same-type constraints will have, compared to rewrite constraints, so it's
+For example, consider `C where .T = .U.V and .U = i32`. The initial lookup for
+`.V` in `.U.V` will fail when checked with `.U` of type `Type`, even though it
+might succeed later if `i32` has a member of type `V`. And worse, for
+`C where .T = F(.Self) and .U = G(.Self)` for arbitrary symbolically-evaluatable
```
```suggestion
`C where .T = F(.Self) and .U = G(.Self)` for arbitrary symbolically evaluatable
```

josh11b commented on proposals/p2173.md:
````diff
@@ -265,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+}
+```
+
+In the case where the value being substituted for `.Self` has rewrites, we need
````
I had trouble understanding from this point.

josh11b commented on proposals/p2173.md:
```diff
@@ -265,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+
+In more detail, what happens here is:
+
+1.  The constraint is fully elaborated. `.X` is replaced with `.Self.X`. For
+    example, `C where .T = .U and .U = i32` is rewritten to
+    `C where .T = .Self.U and .U = i32`.
```
Weird this only happens to the right-hand side of the `=`

zygoloid replied:
The left-hand side is not an expression, it's the name of an associated constant, so this rewrite isn't possible there. But in any case, I felt this description didn't add anything under the new constraint resolution rules, so I removed it.

josh11b commented on proposals/p2173.md:
````diff
@@ -265,20 +276,422 @@ Rewrites aren't performed on the left-hand side of such an `=`, so
+fn BadI[A:! (C where .T = i32) & (C where .U = .T)]() {}
+```
+
+In more detail, what happens here is:
````
Not super clear what example is being referred to by "here"

zygoloid replied:
This text has been removed.

josh11b commented on proposals/p2173.md:
```diff
@@ -294,94 +707,76 @@ interface Example {
-// ❌ Error: applies the same rewrite twice, even
-// though there is no infinite loop.
-fn F4(T:! FalsePositive, x: T.Y.Y.X);
+// ✅ The type of `x` is `T.X.Y.Y`.
```
```suggestion
// ✅ The final type of `x` is `T.X.Y.Y`. It is computed as follows:
```

josh11b commented on proposals/p2173.md:
```diff
@@ -294,94 +707,76 @@ interface Example {
-    `Container & (Iterable where .Element = i32)` is the same as the constraint
-    `(Container & Iterable) where .Element = i32` which is the same as the
-    constraint `Container where .Element = i32`.
+However, some existing forms of non-termination may remain, such as substitution
+into a named constraint recursively triggering another such substitution, or
+template instantiation triggering another template instantiation. Such cases
+will need to be detected and handled in some way, such as by a depth limit, but
```
Do we have actual examples where this happens, or is this just a hedge?

zygoloid replied:
Template instantiation triggering more template instantiation is a real case where I would expect this can happen, just as it does in C++. I think we might have handled the other cases I was thinking about here with your recent constraint completeness proposal, so I've removed mention of that. We also need to finalize the rules for preventing infinite regress when `impl` selection recursively performs more `impl` selection.

zygoloid commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1131 @@
+  fn Make() -> R;
+}
+fn F[T:! X where .R == i32]() -> i32 {
+  return F.Make() + 1;
```
```suggestion
  return T.Make() + 1;
```

zygoloid commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1131 @@
+  let R:! Hashable;
+  fn Make() -> R;
+}
+fn F[T:! X where .R == i32]() -> i32 {
```
```suggestion
fn F[T:! X where .R == i32](x: T) -> i32 {
```

zygoloid commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1131 @@
+class ClassHasX {
+  class X {}
+}
+class HasAssoc {
```
```suggestion
interface HasAssoc {
```

zygoloid commented on proposals/p2173.md:
````diff
@@ -0,0 +1,1131 @@
+their interfaces. This leads to a surprise:
+
+```
+fn G[T:! X where .R == i32]() -> u128 {
````
```suggestion
fn G[T:! X where .R == i32](x: T) -> u128 {
```

zygoloid committed:
Updates based on implementation experience.

zygoloid committed:
Updates based on review feedback. Precommit.

zygoloid committed:
Fix examples

zygoloid reviewed (commented):
@josh11b This has been substantially revised based on implementation experience and a switch to resolving rewrites late rather than immediately. I would appreciate any further feedback!

josh11b commented on proposals/p2173.md:
```diff
@@ -313,49 +317,165 @@ proposal suggests a simpler rule:
-    the result is an error. Substitution into `.Self.U` produces `V.U`, giving
-    an error.
+Within an interface or named constraint, the `impl T as C` syntax permits `=`
+constraints. However, because these constraints don't change the type of `T`,
+rewrite constraints will never result in a rewrite being performed, and instead
+are equivalent to `==` constraints.
```
Could you go into more detail about what goes wrong if they were still treated as rewrites? Is this a case where we don't know to look for the rewrites, we don't know the order in which rewrites should be applied, or something else?

I'm worried since this is an obstacle to eventually simplifying by dropping support for `==` constraints entirely if we can make `=` constraints cover all needed use cases.

zygoloid replied:
The problem is that they're not part of the type of the thing they're rewriting, and only looking in the type of the thing on the left of the `.` for rewrites is the basis for the model and the foundation of the termination argument.

We *could* permit this for `impl as` constraints, but not for `impl SomeOtherType as` constraints, and it seemed more consistent to me to draw the line at the difference between internal (affecting-the-interface-of-`Self`) implementations and external (not-affecting-the-interface-of-`Self`) implementations.

Perhaps we should talk through this more on Discord / open discussion. Maybe there's a way to make this work.

josh11b replied:
Could you add the explanation about `impl SomeOtherType as` to the proposal? I can accept the limitation in the `SomeOtherType != Self` case. I think there is a legitimate question about the `impl as` form. I'm fine with either preliminary decision about what side of the line that form ends up on, but I think we should acknowledge that there are two alternatives under consideration and that we might revisit that specific decision in the future with more experience.

zygoloid replied:
I've added a section to "Alternatives considered" on this.

josh11b commented on proposals/p2173.md:
````diff
@@ -313,49 +317,165 @@ proposal suggests a simpler rule:
+    `.Self` is replaced by `T` throughout the constraint.
+
+```
+// ✅ `.T` is rewritten to `i32` when initially
````
Is `.U` also rewritten?
```suggestion
// ✅ `.T` and `.U` are rewritten to `i32` when initially
```

zygoloid replied:
No, this is specifically referring to the occurrence of `.T` in `.U = .T`.
```suggestion
// ✅ `.T` in `.U = .T` is rewritten to `i32` when initially
```

zygoloid replied:
(My suggestion has been applied; reopening so you can check if this resolves your comment.)

josh11b commented on proposals/p2173.md:
```diff
@@ -708,20 +752,25 @@ interface Example {
-// In `(T.A.A).A`, the inner `T.A.A` is rewritten to `T.A`,
-// resulting in `(T.A).A`, which is then rewritten to `T.A`.
-fn F(T:! Allowed, x: T.A.A.A);
+// ✅ The type of `x` is `T`.
```
```suggestion
// ✅ The final type of `x` is `T`. It is computed as follows:
```

josh11b commented on proposals/p2173.md:
```diff
@@ -883,13 +959,28 @@ understood by developers, and that the more-ergonomic `=` constraints cover
 sufficiently many scenarios that the developer seldom needs to write code to
 request a conversion between types that are known to be the same.
 
+Ideally, we hope that`=` constraints will cover all real situations, and `==`
```
```suggestion
Ideally, we hope that `=` constraints will cover all real situations, and `==`
```

josh11b reviewed (commented):
I think this is really close, I'm very excited!

zygoloid committed:
Apply suggestions from code review

Co-authored-by: josh11b <josh11b@users.noreply.github.com>

zygoloid committed:
Update proposals/p2173.md

zygoloid committed:
Add alternative considered for treatment of `impl as`.

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1291 @@
+        and `==` where constraints, which are similar to, but somewhat different
+        from, those in this proposal.
+    -   `=` constraints require a concrete type on their right-hand side. The
+        left-hand side is not restricted. The type-of-type of the left-hand side
```
Here and elsewhere, I think it would be worthwhile to update this proposal to reflect the new terminology rather than "type-of-type".

josh11b replied:
That seems premature since #2360 hasn't been approved.

chandlerc replied:
Doh, forgot that didn't quite land. Let's try to land that, and then do this? =D

chandlerc replied:
(and this is merged now)

zygoloid replied:
Should be done, please let me know if you find any more spots that are still using the old terminology!

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1291 @@
+## Proposal
+
+Replace the existing `where A == B` constraint with two different kinds of
+constraint:
```
If we're not accepting #2070, shouldn't this *also* replace the existing `=` syntax & semantics?

zygoloid replied:
Done.

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1291 @@
+    `F(T) is SameAs(F(U)`). `observe` statements can be used to form transitive
+    `SameAs` relations. Same-type constraints are not used automatically by the
+    language rules for any purpose, but there is an `ImplicitAs(T)` impl for
+    types implementing `SameAs(T)`.
```
Unless the plan is to actually have a name `SameAs` here, maybe something like (but I'm probably not getting the syntax or phrasing right):

```suggestion
    language rules for any purpose, but there is a blanket impl of `U as ImplicitAs(T) where U == T`.
```

Basically, take the implicit conversion back to `where X == Y` syntax.

zygoloid replied:
Done.

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1291 @@
+This notation is permitted anywhere a constraint can be written, and results in
+a new constraint with a different interface: the named member effectively no
+longer names an associated constant of the constrained type, and is instead
+treated as a rewrite rule the expands to the right-hand side of the constraint,
```
```suggestion
treated as a rewrite rule that expands to the right-hand side of the constraint,
```

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1291 @@
+`C where .R = i32` don't differ by merely having more type information; rather,
+they are different types that have an isomorphic set of values, somewhat like
+`i32` and `u32`. An `=` constraint is not merely learning a new fact about a
+type, it is requesting different behavior.
```
I suspect this section should also be updated w/ the new terminology even though it doesn't mention "type-of-type" directly?

zygoloid replied:
Done.

chandlerc commented on proposals/p2173.md:
````diff
@@ -0,0 +1,1291 @@
+fn F(T:! B) -> T.(A.T) { return n; }
+```
+
+### Combining constraints with `impl as`
+
+Within an interface or named constraint, the `impl T as C` syntax permits `=`
+constraints. However, because these constraints don't change the type of `T`,
+rewrite constraints will never result in a rewrite being performed, and instead
+are equivalent to `==` constraints.
+
+```
+interface A {
+  let T:! Type;
+  let U:! Type;
+}
+constraint C {
+  extends A where .T = .U and .U = i32;
+}
+constraint D {
+  extends A where .T == .U and .U == i32;
+}
+interface B {
+  // This constraint is equivalent to
+  // `impl as A where .T == .U and .U == i32;` or
+  // `impl as C;` or `impl as D;`.
+  impl as A where .T = .U and .U = i32;
+}
+
+var n: i32;
+
+// ❌ No implicit conversion from `i32` to `T.(A.T)`.
+// Resolved constraint on `T` is
+// `B where T.(A.T) == T.(A.U) and T.(A.U) = i32`.
+// `T.(A.T)` is single-step equal to `T.(A.U)`, and
+// `T.(A.U)` is single-step equal to `i32`, but
+// `T.(A.T)` is not single-step equal to `i32`.
+fn F(T:! B) -> T.(A.T) { return n; }
+```
````
When I first read this, I thought it also applied to the case of actually introducing an implementation, which didn't make sense. I actually wrote up a bunch of comments here about this, that were all better covered by the section below. =D

Could you start this section off with a clarification that this only applies to introducing a requirement for an `impl` in another interface or constraint? Maybe worth a forward reference as well for the other cases.

---

Is the reason to allow writing rewrite-constraint *syntax* here because they are allowed (and meaningful) in `extends`? Could we remove that easily, or would that significantly regress the ergonomics?

I would lean towards disallowing the writing of rewrite-constraints in contexts where they're not going to apply. Even if that means that "laundering" a rewrite constraints from an extending constraint into an `impl as` can still surface the behavior, that doesn't seem as likely to be encountered.

What do others think?

josh11b replied:
This is perhaps an argument to select the "Rewrite constraints in `impl as` constraints" alternative -- so there isn't a cliff between `extends` and `impl as`. Then we might have this restriction on rewrite constraints only on `impl OtherType as` declarations.

chandlerc replied:
> This is perhaps an argument to select the "Rewrite constraints in `impl as` constraints" alternative -- so there isn't a cliff between `extends` and `impl as`. Then we might have this restriction on rewrite constraints only on `impl OtherType as` declarations.

Yeah, that would make sense to me.

Or more generally, preclude rewrite constraints when the type can be something other than Self -- whether it is *spelled* `impl as` or `impl T as` maybe isn't as important as whether we can prove that `T` is exactly `Self`. (If I understood the necessary part correctly?)

My preference to *completely* reject them (rather than allow the syntax but w/o the rewrite semantic) remains though.

zygoloid replied:
I think the biggest issue here is rewrite constraints that are found _indirectly_:

```
interface A { let T:! type; }
constraint B { extends C where .T = i32; }
interface C { impl as B; }
```

Here, I think it's probably important that we accept the `impl as B`, even though `B` includes a rewrite constraint that won't end up being part of `C`. Rejecting this case seems likely to be problematic -- it'd make `=` constraints and `impl as` incompatible. As an example, the presence of an `impl as` in a consumer of a constraint would prevent certain forms of library evolution, such as adding a new associated constant with a default value, along with a rewrite constraint for that associated constant in some derived constraint.

We could try to add a _syntactic_ restriction on `where ... = ...` appearing at the top level within an `impl as`. Some choices here, such as rejecting any `=` constraint is written that are written anywhere within an `impl as` declaration, or tracking the provenance of the constraint in some way, seem awkward and somewhat arbitrary. But perhaps we could use a similar rule to the one we picked for scoping / reuse of `.Self`, and only recurse into `where` and `&` expressions and parentheses on the RHS of `impl ? as`, and reject any `=` constraints we find there. I think I'd be OK with that, but we should discuss whether we think that check is better implemented as a language rule or a style rule.

#2476 is also somewhat relevant here.

josh11b replied:
Is that supposed to be `constraint B { extends A where .T = i32; }`?

zygoloid replied:
Oops, yes it is :)

zygoloid replied:
I'm sufficiently convinced that a syntactic restriction makes sense here that I've added one to the proposal.

josh11b replied:
To summarize my understanding of the discussion so far, given this example:
```
interface A { let T:! type; }
constraint B { extends A where .T = i32; }
interface C { impl as B; }
class D { impl as C {} }
interface E { impl as A where .T = i32; }
interface F(U:! Type) { impl U as B; }
```
We have listed the following options:
- Proposal as-is: the above is accepted with `=` constraints changed to `==` constraints in interfaces `C`, `E`, and `F`. This means `D.(A.T) == i32` but there is no rewrite `D.(A.T)` ~> `i32`.
- We never replace `=` constraints into `==` constraints, but that means rejecting interfaces `C`, `E`, and `F` above.
- We reject `=` constraints in a `impl as __ where ...` syntactically, as in `interface E`, but allow replacing them with `==` constraints when used indirectly as in interfaces `C` and `F`.
- We adopt the "Rewrite constraints in `impl as` constraints" alternative, so interfaces `C` and `E` are accepted and those `=` constraints are consistently rewrite constraints. We then have to select one of the above options for interface `F`.

zygoloid replied:
I think that's right, but the proposal as-is is now your third option. I realized the same thing also happens for `where ... is (__ where ...)`, so I've called that case out too; I think the same rules should apply there, because `where T is C` is the `where` equivalent of `impl T as C`.

chandlerc replied:
I'm happy with this end state.

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1291 @@
+    rule cannot be readily understood by a developer nor effectively worked
+    around.
+-   Rust also has an
+    [undecidable type system (content warning: swear words)](https://sdleffler.github.io/RustTypeSystemTuringComplete/)
```
Trivial nit: slightly more precise / clear CW?

```suggestion
    [undecidable type system (content warning: profanity)](https://sdleffler.github.io/RustTypeSystemTuringComplete/)
```

Also, wow, this is such a rabbit hole of esoteric languages...

zygoloid replied:
I'd like to push back on this suggestion just a little: in this context, the word in question is not being used as profanity but simply as part of a proper noun, but it is nonetheless objectively a swear word. Happy to accept the suggestion if you still think it's a beneficial change given that argument.

chandlerc replied:
I see your point there.

I struggle with "swear words" because for me it seems like a slang term or idiom and doesn't unambiguously communicate the severity of the word choice. A result is that I'm not sure it'll serve its purpose reliably.

Maybe "obscenities" or "obscene words" or "obscene language" would work to avoid talking about the *usage* but still make the warning a bit more pointed?

zygoloid replied:
I've tried a rewrite to make this clearer, at the expense of being a lot more verbose :)

chandlerc reviewed (changes requested):
Two real things I think are still needed here:

1) Updating to the new terminology direction, and
2) Clarifying (and maybe tweaking) the `impl as` constraints.

I'm suggesting (1) here even in the proposal because I suspect this particular proposal and write-up will end up fairly heavily read and cited, so it seems worthwhile to align it to the newer terminology and stuff like `type` to avoid confusion in the future.

zygoloid committed:
Update terminology to match #2360 and address some minor review
comments.

zygoloid committed:
Mention early on that we're restricting the syntax of `where A = B`, not
only changing the semantics.

zygoloid committed:
Disallow top-level rewrites in `impl as` and also in `is` constraints.

chandlerc commented:
IS this ready for another round of review? (It wasn't marked as such, but looks like it might be...)

zygoloid commented:
> IS this ready for another round of review? (It wasn't marked as such, but looks like it might be...)

Yes. What kind of marking are you looking for?

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1355 @@
+
+It's important to distinguish here between learning more about a type, such as
+learning that it implements a particular interface – for example, from an added
+import – and changing the type, such as by writing a different expression after
+a `:` or `:!`. Changing the type of a value is expected to be able to change the
```
IIUC, with the new terminology, a different expression in type position doesn't *change* the type of the value any more? Guessing this is just another place needing a terminology update.

zygoloid replied:
What I mean here is that changing from `a: i32` to `a: f32` changes the type of `a`, and changing from `T:! C` to `T:! C where .X = 5` changes the type of `T`. I think what you're observing is that changing from `T:! C = i32` to `T:! (C where .X = 5) = i32` doesn't change the type of `a: T` -- either way, it's `a: i32`, because the right-hand side of `:` is converted to `type`.

Is there some way I can update this paragraph to make it clearer what I'm talking about?

chandlerc replied:
I can't think of anything.

The root issue is "writing a different expression after a `:` or `:!`." -> we've just recently allowed that to *not* always change the type, so I get a little surprised when that is the first example of how to change the type of the value. But of course, it is, and different expressions that *don't* change the type of the value are the exception and not the rule. So I think its ultimately fine. =D

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1355 @@
+Within an interface or named constraint, the `impl T as C` syntax does not
+permit `=` constraints to be specified directly. However, such constraints can
+be specified indirectly, for example if `C` is a named constraint that contains
+rewrites. Because these constraints don't change the type of `T`, rewrite
```
Not really about this review, but mostly a question I had when reading this and wanted to write down so I didn't forget it or forget to ask about it. Happy to just discuss this directly to deepen my understanding of how this stuff works.

I'm tempted to build up a mental model for the restrictions in this space something like the following....

We don't propagate rewrite rules here because `impl T as C` or an `is` constraint like `... where .T is C` doesn't actually change the type (or even facet) `T` in any way. Instead, these merely establish that `T` is a value that *could* be used with the facet type `C`. The rewrite applies to `C`, but that has no bearing on `T`. In a sense, it vanishes. Am I good so far?

(Less sure about this part) This is essence the same distinction between `... where T = U` and `... where T == U`, but instead of between two specific facets, its between a facet and a facet type. The first (rewrite constraints) is analogous to `let T:! C` where rewrites in `C` apply to associated constants of `T`, but in `let T:! ... where .Self is C` (if I've spelled that right) they would *not*. Is that a reasonable way to be thinking about these things?

And then the syntactic restriction is just to preclude writing a rewrite constraint in a syntactic context where it *couldn't* have meaning. From above, `C` can be used either in a way where the rewrite applies or where it doesn't. But when the rewrite only exists directly within a context that cannot propagate it, it would have no meaning, and we don't need to let you pointless write a spelling without (additional to `==`) meaning.

Understanding all of this in this way -- that `is` and `impl T as` are really establishing a *satisfaction* of a constraint, not *applying* it (in the same way that `... where T == U` establishes a *satisfaction* of same-type, but doesn't apply that by rewriting) helps me rationalize and reason about the boundary across which rewrites don't flow. But wanted to write all this down in different words and run it by you all just as a check of my understanding.

Even if everything here checks out, I have no reason to suspect a different write-up would be specifically beneficial at this stage. I'm just trying to make sure I'm learning and internalizing the model correctly. =D

zygoloid replied:
Yes, that all looks correct to me.

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1355 @@
+
+-   When the constraint is used explicitly, when declaring a generic parameter
+    or associated constant of the form `T:! Constraint`.
+-   When declaring an `impl` such as `impl T as Constraint`.
```
This is when declaring an `impl` *outside* of an interface or named constraint?

zygoloid replied:
Yes. I've made this more explicit.

chandlerc commented on proposals/p2173.md:
```diff
@@ -0,0 +1,1355 @@
+new recursive type-checking steps, so should not introduce any new forms of
+non-termination. Rewrite constraints thereby give us a deterministic,
+terminating type canonicalization mechanism for associated constants: in `A.B`,
+if the type of `A` specifies that `.B == C`, then `A.B` is replaced by `C`.
```
Er, shouldn't this be `.B = C`? Or did I miss something?

```suggestion
if the type of `A` specifies that `.B = C`, then `A.B` is replaced by `C`.
```

zygoloid replied:
Oops! Thanks!

chandlerc reviewed (approved):
I think this is in great shape. One comment below is just a discussion topic I wrote down to avoid forgetting. The rest I think are super minor. If they all make sense, don't block, but feel free to ping for a quick look. I don't tihnk anything left needs re-reading significant sections.

Also, this is a fantastic write-up and I'm really excited about the whole thing. It's a remarkably complex problem and area, and the treatment here I think is great.

josh11b commented on proposals/p2173.md:
```diff
@@ -396,36 +404,83 @@ var n: i32;
+// ❌ Rewrite constraint specified directly in `is`.
+fn F[T:! A where .U is (A where .T = i32)]();
+
+// ❌ Reference to `.U` in same-type constraint is ambiguous:
+// does this mean the outer or inner `.Self.U`?
```
Do you mean `.T` instead of `.U`?

zygoloid replied:
Yes, thanks, fixed.

josh11b commented on proposals/p2173.md:
```diff
@@ -396,36 +404,83 @@ var n: i32;
+// ✅ Not specified directly, but does not result
+// in any rewrites being performed. Return type
+// is not rewritten to `i32`.
+fn H[T:! type where T is C]() -> T.(A.U);
```
I don't think the name `T` is available until the end of the binding:
```suggestion
fn H[T:! type where .Self is C]() -> T.(A.U);
```

zygoloid replied:
Fixed.

zygoloid committed:
Address review comments.

chandlerc reviewed (approved):
Ship it!
